src/app/models/auth_source_cas.rb

```diff
@@ -58,37 +58,115 @@ def add_user_to_group(groupname, user)
 
   def api_request(uri, form_data)
     http_uri = URI.parse(uri)
-    http = Net::HTTP.new(http_uri.host,http_uri.port)
+    http = Net::HTTP.new(http_uri.host, http_uri.port)
     http.use_ssl = true
     http.verify_mode = OpenSSL::SSL::VERIFY_NONE
-    request = Net::HTTP::Post.new(http_uri.path, initheader = {'Content-Type' =>'application/json'})
+    request = Net::HTTP::Post.new(http_uri.path, initheader = { 'Content-Type' => 'application/json' })
     request.set_form_data(form_data)
     return http.request(request)
   end
 
+  def self.create_or_update_user(login, user_givenName, user_surname, user_mail, user_groups, auth_source_id)
+    # Get ces admin group
+    admingroup_exists = false
+    if Ces_admin_group != ''
+      admingroup_exists = true
+    end
+
+    user = User.find_by_login(login)
+    if user == nil # user not in redmine yet
+
+      user = User.new
+      user.login = login
+      user.firstname = user_givenName
+      user.lastname = user_surname
+      user.mail = user_mail
+      user.auth_source_id = auth_source_id
+      if admingroup_exists
+        if user_groups.to_s.include?(Ces_admin_group.gsub('\n', ''))
+          user.admin = 1
+        end
+      end
+
+      for i in user_groups
+        # create group / add user to group
+        AuthSourceCas.add_user_to_group(i.content.to_s, user)
+      end
+
+      if !user.save
+        raise user.errors.full_messages.to_s
+      end
+    else
+      # user already in redmine
+      @usergroups = Array.new
+      for i in user_groups
+        @usergroups << i.content.to_s
+        # create group / add user to group
+        add_user_to_group(i.content.to_s, user)
+      end
+      # remove user from groups he is not in any more
+      @casgroups = Group.where(firstname: 'cas')
+      for l in @casgroups
+        @casgroup = Group.find_by(lastname: l.to_s)
+        @casgroupusers = User.active.in_group(@casgroup).all()
+        for m in @casgroupusers
+          if (m.login == login) and not (@usergroups.include?(l.to_s))
+            @casgroup.users.delete(user)
+          end
+        end
+      end
+      # remove user's admin rights if he is not in admin group any more
+      casAdminPermissionsCustomField = UserCustomField.find_by_name('casAdmin')
+      # We currently save the value for the casAdmin Field as `true` or `false`. However, redmine saves them as `1` and `0`. We need to support both.
+      wasCreatedByCAS = user.custom_field_value(casAdminPermissionsCustomField).is_true?
+      if admingroup_exists and wasCreatedByCAS
+        if user_groups.to_s.include?(Ces_admin_group.gsub('\n', ''))
+          user.admin = 1
+        else
+          user.admin = 0
+        end
+        user.save
+      end
+    end
+
+    return user
+  end
+
   def authenticate(login, password)
     return nil if login.blank? || password.blank?
 
+    user = User.find_by_login(login)
+    if user.api_key === password
```

Comment:
This isn't how the API keys should work according to the redmine documentation: https://www.redmine.org/projects/redmine/wiki/rest_api#Authentication